File: sdks/php/examples/basic_usage.php
<?php

require_once __DIR__ . '/../vendor/autoload.php';

use Toonstore\Torm\TormClient;

echo "🚀 TORM PHP SDK - Basic Usage Example\n\n";

// 1. Connect to TORM server
echo "Connecting to TORM server...\n";
$torm = new TormClient('http://localhost:3001');

try {
    $health = $torm->health();
    echo "✅ Connected! Status: " . ($health['status'] ?? 'unknown') . "\n\n";
} catch (Exception $e) {
    echo "❌ Failed to connect: {$e->getMessage()}\n";
    exit(1);
}

// 2. Define User model with validation
echo "Defining User model...\n";
$User = $torm->model('User', [
    'name' => ['type' => 'string', 'required' => true, 'min_length' => 3],
    'email' => ['type' => 'string', 'required' => true, 'email' => true],
    'age' => ['type' => 'integer', 'min' => 13, 'max' => 120],
    'active' => ['type' => 'boolean']
]);
echo "✅ User model defined\n\n";

// 3. Create users
echo "Creating users...\n";
try {
    $alice = $User->create([
        'id' => 'user:alice',
        'name' => 'Alice Smith',
        'email' => 'alice@example.com',
        'age' => 30,
        'active' => true
    ]);
    echo "✅ Created: " . $alice['name'] . "\n";
    
    $bob = $User->create([
        'id' => 'user:bob',
        'name' => 'Bob Johnson',
        'email' => 'bob@example.com',
        'age' => 25,
        'active' => true
    ]);
    echo "✅ Created: " . $bob['name'] . "\n";
    
    $charlie = $User->create([
        'id' => 'user:charlie',
        'name' => 'Charlie Brown',
        'email' => 'charlie@example.com',
        'age' => 35,
        'active' => false
    ]);
    echo "✅ Created: " . $charlie['name'] . "\n\n";
} catch (Exception $e) {
    echo "❌ Failed to create user: {$e->getMessage()}\n\n";
}

// 4. Find all users
echo "Finding all users...\n";
try {
    $allUsers = $User->find();
    echo "✅ Found " . count($allUsers) . " users\n";
    foreach ($allUsers as $user) {
        echo "   - {$user['name']} ({$user['email']})\n";
    }
    echo "\n";
} catch (Exception $e) {
    echo "❌ Failed to find users: {$e->getMessage()}\n\n";
}

// 5. Find user by ID
echo "Finding user by ID...\n";
try {
    $user = $User->findById('user:alice');
    if ($user) {
        echo "✅ Found: " . $user['name'] . "\n\n";
    } else {
        echo "❌ User not found\n\n";
    }
} catch (Exception $e) {
    echo "❌ Failed to find user: {$e->getMessage()}\n\n";
}

// 6. Query with filters
echo "Querying active users over 25...\n";
try {
    $results = $User->query()
        ->filter('active', 'eq', true)
        ->filter('age', 'gte', 25)
        ->sort('age', 'asc')
        ->exec();
    
    echo "✅ Found " . count($results) . " matching users:\n";
    foreach ($results as $user) {
        echo "   - {$user['name']}, age {$user['age']}\n";
    }
    echo "\n";
} catch (Exception $e) {
    echo "❌ Query failed: {$e->getMessage()}\n\n";
}

// 7. Update user
echo "Updating user...\n";
try {
    $updated = $User->update('user:bob', ['age' => 26]);
    echo "✅ Updated: {$updated['name']}, new age: {$updated['age']}\n\n";
} catch (Exception $e) {
    echo "❌ Failed to update user: {$e->getMessage()}\n\n";
}

// 8. Count users
echo "Counting users...\n";
try {
    $count = $User->count();
    echo "✅ Total users: $count\n\n";
} catch (Exception $e) {
    echo "❌ Failed to count users: {$e->getMessage()}\n\n";
}

// 9. Validation demo
echo "Testing validation...\n";
try {
    $User->create([
        'id' => 'user:invalid',
        'name' => 'Invalid User',
        'email' => 'not-an-email',
        'age' => 30
    ]);
    echo "❌ Validation didn't catch invalid email\n\n";
} catch (Exception $e) {
    echo "✅ Validation caught error: {$e->getMessage()}\n\n";
}

// 10. Delete user
echo "Deleting user...\n";
try {
    $success = $User->delete('user:charlie');
    if ($success) {
        echo "✅ User deleted successfully\n\n";
    } else {
        echo "❌ Failed to delete user\n\n";
    }
} catch (Exception $e) {
    echo "❌ Failed to delete user: {$e->getMessage()}\n\n";
}

// 11. Verify deletion
echo "Verifying deletion...\n";
try {
    $user = $User->findById('user:charlie');
    if ($user === null) {
        echo "✅ User successfully deleted\n\n";
    } else {
        echo "❌ User still exists\n\n";
    }
} catch (Exception $e) {
    echo "❌ Failed to verify: {$e->getMessage()}\n\n";
}

echo "🎉 Example completed!\n";

File: sdks/python/examples/basic_usage.py
"""
TORM Python SDK - Basic Usage Example

This example demonstrates how to use the ToonStore ORM Python client.
"""

from toonstore_torm import TormClient

def main():
    print("🚀 TORM Python SDK - Basic Usage Example\n")
    
    # 1. Connect to TORM server
    print("Connecting to TORM server...")
    torm = TormClient('http://localhost:3001')
    
    # Check health
    try:
        health = torm.health()
        print(f"✅ Connected! Status: {health.get('status', 'unknown')}\n")
    except Exception as e:
        print(f"❌ Failed to connect: {e}")
        return
    
    # 2. Define User model with validation
    print("Defining User model...")
    User = torm.model('User', {
        'name': {'type': 'str', 'required': True, 'min_length': 3},
        'email': {'type': 'str', 'required': True, 'email': True},
        'age': {'type': 'int', 'min': 13, 'max': 120},
        'active': {'type': 'bool'}
    })
    print("✅ User model defined\n")
    
    # 3. Create users
    print("Creating users...")
    try:
        alice = User.create({
            'id': 'user:alice',
            'name': 'Alice Smith',
            'email': 'alice@example.com',
            'age': 30,
            'active': True
        })
        print(f"✅ Created: {alice.get('name')}")
        
        bob = User.create({
            'id': 'user:bob',
            'name': 'Bob Johnson',
            'email': 'bob@example.com',
            'age': 25,
            'active': True
        })
        print(f"✅ Created: {bob.get('name')}")
        
        charlie = User.create({
            'id': 'user:charlie',
            'name': 'Charlie Brown',
            'email': 'charlie@example.com',
            'age': 35,
            'active': False
        })
        print(f"✅ Created: {charlie.get('name')}\n")
    except Exception as e:
        print(f"❌ Failed to create user: {e}\n")
    
    # 4. Find all users
    print("Finding all users...")
    try:
        all_users = User.find()
        print(f"✅ Found {len(all_users)} users")
        for user in all_users:
            print(f"   - {user.get('name')} ({user.get('email')})")
        print()
    except Exception as e:
        print(f"❌ Failed to find users: {e}\n")
    
    # 5. Find user by ID
    print("Finding user by ID...")
    try:
        user = User.find_by_id('user:alice')
        if user:
            print(f"✅ Found: {user.get('name')}\n")
        else:
            print("❌ User not found\n")
    except Exception as e:
        print(f"❌ Failed to find user: {e}\n")
    
    # 6. Query with filters
    print("Querying active users over 25...")
    try:
        results = User.query() \
            .filter('active', 'eq', True) \
            .filter('age', 'gte', 25) \
            .sort('age', 'asc') \
            .exec()
        
        print(f"✅ Found {len(results)} matching users:")
        for user in results:
            print(f"   - {user.get('name')}, age {user.get('age')}")
        print()
    except Exception as e:
        print(f"❌ Query failed: {e}\n")
    
    # 7. Update user
    print("Updating user...")
    try:
        updated = User.update('user:bob', {'age': 26})
        print(f"✅ Updated: {updated.get('name')}, new age: {updated.get('age')}\n")
    except Exception as e:
        print(f"❌ Failed to update user: {e}\n")
    
    # 8. Count users
    print("Counting users...")
    try:
        count = User.count()
        print(f"✅ Total users: {count}\n")
    except Exception as e:
        print(f"❌ Failed to count users: {e}\n")
    
    # 9. Validation demo
    print("Testing validation...")
    try:
        # This should fail - email invalid
        User.create({
            'id': 'user:invalid',
            'name': 'Invalid User',
            'email': 'not-an-email',
            'age': 30
        })
        print("❌ Validation didn't catch invalid email\n")
    except Exception as e:
        print(f"✅ Validation caught error: {e}\n")
    
    # 10. Delete user
    print("Deleting user...")
    try:
        success = User.delete('user:charlie')
        if success:
            print("✅ User deleted successfully\n")
        else:
            print("❌ Failed to delete user\n")
    except Exception as e:
        print(f"❌ Failed to delete user: {e}\n")
    
    # 11. Verify deletion
    print("Verifying deletion...")
    try:
        user = User.find_by_id('user:charlie')
        if user is None:
            print("✅ User successfully deleted\n")
        else:
            print("❌ User still exists\n")
    except Exception as e:
        print(f"❌ Failed to verify: {e}\n")
    
    # Close connection
    torm.close()
    print("🎉 Example completed!")


if __name__ == '__main__':
    main()
